keras_nlp/models/whisper/whisper_cached_multi_head_attention.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Whisper Cached Multi-Head Attention layer."""
+
+
+import string
+
+import tensorflow as tf
+from keras.layers import core
+from keras.utils import tf_utils
+
+from keras_nlp.layers.cached_multi_head_attention import (
+    CachedMultiHeadAttention,
+)
+
+_CHR_IDX = string.ascii_lowercase
+
+
+def _build_proj_equation(free_dims, bound_dims, output_dims):
+    """Builds an einsum equation for projections inside multi-head attention."""
+    input_str = ""
+    kernel_str = ""
+    output_str = ""
+    bias_axes = ""
+    letter_offset = 0
+    for i in range(free_dims):
+        char = _CHR_IDX[i + letter_offset]
+        input_str += char
+        output_str += char
+
+    letter_offset += free_dims
+    for i in range(bound_dims):
+        char = _CHR_IDX[i + letter_offset]
+        input_str += char
+        kernel_str += char
+
+    letter_offset += bound_dims
+    for i in range(output_dims):
+        char = _CHR_IDX[i + letter_offset]
+        kernel_str += char
+        output_str += char
+        bias_axes += char
+    equation = f"{input_str},{kernel_str}->{output_str}"
+
+    return equation, bias_axes, len(output_str)
+
+
+def _get_output_shape(output_rank, known_last_dims):
+    return [None] * (output_rank - len(known_last_dims)) + list(known_last_dims)
+
+
+class WhisperCachedMultiHeadAttention(CachedMultiHeadAttention):
+    """Whisper Cached Multi-Head Attention layer.
+
+    Inherits from `keras_nlp.layers.cached_multi_head_attention.CachedMultiHeadAttention`,
```

Comment:
So the only thing we need here is to avoid the `K` bias term in that dense layer? I wonder if we could be a little hack on the whisper encoder and whisper decoder layers, and just do something like...

```python
class WhisperEncoder(TransformerEncoder):
    def _build(self, input_shape):
        super()._build(input_shape)
        # Whisper encoder is almost standard but has no bias for the key projection.
        # For simplicity of code, we just leave the key layer's bias zero and untrainable.
        self._self_attention_layer._key_dense.bias.trainable = False
```

Could something like this work you think?

Reply:
Wow, very nice! It should work, yeah

Reply:
Trying it out